### PR TITLE
Infer NumPyro InferenceData dims automatically

### PR DESCRIPTION
**Pull request by kylejcaron**
## Description

Fixes: #2022

 If dims are not provided, `az.from_numpyro` now infers dims automatically based on model plates, sample sites, and provided `coords`. 

For event_dims...
 * for numpyro.sample sites, infer={"event_dims":["dim1"]} can be provided
 * for deterministic sites, there is an `extra_event_dims` field in az.from_numpyro to manually provide them.

`pred_dims` are also now inferred by default if they arent provided.

## Checklist

- [x] Follows [official](https://github.com/arviz-devs/arviz/blob/main/CONTRIBUTING.md#pull-request-checklist) PR format
- [ ] Includes a sample plot to visually illustrate the changes (only for plot-related functions)
- [x] New features are properly documented (with an example if appropriate)?
- [ ] Includes new or updated tests to cover the new feature
- [x] Code style  correct (follows pylint and black guidelines)
- [x] Changes are listed in [changelog](https://github.com/arviz-devs/arviz/blob/main/CHANGELOG.md#v0xx-unreleased)

----
📚 Documentation preview 📚: https://arviz--2441.org.readthedocs.build/en/2441/

### CODE REVIEW

File: arviz/tests/external_tests/test_data_numpyro.py [hunk 1]
```diff
@@ -282,3 +290,61 @@ def model():
...
+        import numpyro.distributions as dist
+        from numpyro.infer import MCMC, NUTS
+        import jax.numpy as jnp
+
+        def model():
+            alpha = numpyro.sample("alpha", dist.Normal(0, 1))
+            sigma = numpyro.sample("sigma", dist.HalfNormal(1))
+            with numpyro.plate("obs_idx", 3):
+                # mu is plated by obs_idx, but isnt broadcasted to the plate shape
+                mu = numpyro.deterministic("mu", alpha)
```
[the remark this answers is not in the document]
Comment (kylejcaron) [reply arviz/tests/external_tests/test_data_numpyro.py:322]: I decided to proceed with the least assumptions, which was option 3. In this example, the dims for alpha simply dont match the plate (even though it could be broadcasted). changed the test to xfail

**Comment by kylejcaron** [comment]
This is ready to be reviewed! 

There are some CI failures that seem unrelated to these changes (Im seeing the same issue on other PRs)

**Review by OriolAbril — COMMENTED** [review]
Thanks for the PR! This will be a very nice addition.

CI should be fixed now, sorry for the slow response, we do what we can.

I looked a bit over the code and if I understand correctly, dimension names will only be inferred when the `coords` argument is explicitly provided but I think it is not necessary to require `coords` (both quickly going over the code and conceptually).

In xarray, dimensions are always named, even if it is a `var_name_id` name, but coordinate values are completely optional to the point that even when no coord values are present `.sel` still works without problem and simply behaves like `.isel`. Consequently, in my opinion, for a plated model where no coords are provided it is preferable to end up with properly named dimensions without coord values than it is to get nothing at all.

In practice I think this will only mean removing the `@requires("coords")` decorators and updating some of the tests to cover that case.

---

Lastly, specially now that the converter has more functionality, I think it would be great to have a page like https://python.arviz.org/en/stable/getting_started/ConversionGuideEmcee.html for numpyro too. This is related to https://github.com/arviz-devs/arviz/issues/1489 and it is not necessary to add the notebook in this PR but it would be great if you could then create a dedicated issue for the numpyro conversion guide with a proposed outline as you probably have a better idea of what to include than me for example. I think this will make the issue much more approachable to other less experienced contributors.

File: arviz/tests/external_tests/test_data_numpyro.py [hunk 2]
```diff
@@ -85,6 +85,7 @@ def test_inference_data_namedtuple(self, data):
         data.obj.get_samples = lambda *args, **kwargs: data_namedtuple
         inference_data = from_numpyro(
             posterior=data.obj,
+            dims={},  # This mock test needs to turn off autodims like so or mock group_by_chain
```
Comment (kylejcaron) [inline arviz/tests/external_tests/test_data_numpyro.py:88]: Calling out this change to a pre-existing test. Looks like my infer_dims logic broke this test, however I found out that its also due to the fact that the mock object for get_samples() didnt have any `group_by_chain` logic. This was causing the test to have the wrong posterior shape of (chain*draw, J) instead of (chain, draw, J). 

the infer_dims logic was catching this and failing, so i chose to turn off the infer_dims logic and let the test work as originally intended (I'm not sure what the original purpose of this test actually is and which edge case its covering)

File: arviz/data/io_numpyro.py [hunk 3]
```diff
@@ -410,14 +409,9 @@ def infer_dims(self) -> Dict[str, List[str]]:
 
     @requires("posterior")
     @requires("model")
-    @requires("coords")
     @requires("predictions")
     def infer_pred_dims(self) -> Dict[str, List[str]]:
-        if self.predictions_constant_data is not None:
-            dims = infer_dims(self.model, model_kwargs=self.predictions_constant_data)
```
Comment (kylejcaron) [inline arviz/data/io_numpyro.py]: Assumption: removed this logic and chose to infer dims with the original input data in all cases - otherwise the `test_inference_data_constant_data` test was failing since it passed in `x2` as prediction constant data, even though x2 isnt a model input

**Comment by kylejcaron** [comment]
Thanks for taking a look! 

> Lastly, specially now that the converter has more functionality, I think it would be great to have a page like https://python.arviz.org/en/stable/getting_started/ConversionGuideEmcee.html for numpyro too. This is related to #1489 and it is not necessary to add the notebook in this PR but it would be great if you could then create a dedicated issue for the numpyro conversion guide with a proposed outline as you probably have a better idea of what to include than me for example. I think this will make the issue much more approachable to other less experienced contributors.

Happy to do so, I'll get started on making the issue (and add in the conversion guide itself if I have time!). 

As part of this, what are your thoughts on also adding numpyro converter logic for variational inference? I know PyMC has one (where chains=1), but I also know its slightly misleading for less experienced users since many of ArviZ's diagnostics are MCMC based. I personally still use inferencedata objects for numpyro SVI results so it would atleast be helpful for myself


> In practice I think this will only mean removing the `@requires("coords")` decorators and updating some of the tests to cover that case.

good callout, I removed that requirement. This lead me to making two bigger assumptions in my logic to get tests passing, I called them out with comments if you want to take a look

File: arviz/tests/external_tests/test_data_numpyro.py (hunk 2 above)
Comment (OriolAbril) [reply arviz/tests/external_tests/test_data_numpyro.py:88]: Also not sure about the goal of the test, it might even predate `group_bu_chain` logic and never gotten updated. It is ok with me to leave like this

**Review by OriolAbril — APPROVED** [review]
